### PR TITLE
Backend Setup

### LINKED ISSUE #692: 🚀 [Feature] Apollo Backend
## 🚀 Feature

Implement the new apollo server backend

### PR DESCRIPTION
**Pull request by ulfgebhardt**
## 🍰 Pullrequest

Backend Setup.

This PR is the first step for a nodejs based backend server. This PR can be merged as soon as the communication between client and server is completely covered by the new backend. Furthermore authentication via JWT would be a good feature to start with

#### TechStack
- [x] typescript
- [x] express
- [x] express-graphql
- [x] eslint
- [x] prettier
- [x] ~~jest~~
- [x] publish workflow
- [x] lint workflow
- [x] ~~test workflow~~
- [x] docker

### Issues

- fixes #692 
- relates #13
- relates #17 
- relates #477 

### Todo

- [x] ~~Implement jest~~
- [ ] Make Issue for jest and test workflow
- [x] Implement docker
- [x] Implement docker-compose setup
- [x] Implement workflows
- [x] ~~Implement typeorm connection~~
- [ ] Make Issue for typeorm implementation
- [ ] Implement proper graphql schema
- [ ] Issue: API Versioning?
- [x] Config (enable graphiql, debug, which api version is supported, port)
- [x] Implement API connection to Login Server
- [ ] Make Issue JWT based on old session
- [x] Remove Book examples

### CODE REVIEW

File: backend/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  extends: [
+    'standard',
+    'eslint:recommended',
+    'plugin:prettier/recommended',
+    'plugin:@typescript-eslint/recommended',
+  ],
+  // add your custom rules here
+  rules: {
+    'no-console': ['error'],
+    'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off',
```
Comment (Mogge) [inline backend/.eslintrc.js:18]: I suggest always error.

File: backend/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+/node_modules/
```
Comment (Mogge) [inline backend/.gitignore:1]: Please add `*~` for my emacs backup files

File: backend/.prettierrc.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  semi: false,
+  printWidth: 100,
+  singleQuote: true,
+  trailingComma: "all",
+  tabWidth: 2,
+  bracketSpacing: true
```
Comment (Mogge) [inline backend/.prettierrc.js:7]: trailing comma all? Did you run the linting here?

File: backend/ormconfig.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "type": "sqlite",
```
Comment (Mogge) [inline backend/ormconfig.json:2]: I guess this is just a temporal solution.

File: backend/package.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "gradido-backend",
+  "version": "0.0.1",
+  "description": "Gradido unified backend providing an API-Service for Gradido Transactions",
+  "main": "index.js",
```
Comment (Mogge) [inline backend/package.json:5]: src/index.js

**Comment by einhornimmond** [comment]
Try Login User, but I get this error:
```
Cannot find module 'class-validator'\nRequire stack:\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\resolvers\\validate-arg.js\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\resolvers\\helpers.js\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\resolvers\\create.js\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\schema\\schema-generator.js\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\utils\\buildSchema.js\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\utils\\index.js\n- F:\\Gradido\\gradido_apollo\\backend\\node_modules\\type-graphql\\dist\\index.js\n- F:\\Gradido\\gradido_apollo\\backend\\src\\index.ts
```

File: backend/.eslintrc.js (hunk 1 above)
Comment (ulfgebhardt) [reply backend/.eslintrc.js:18]: Ok, why do we have this in the frontend? Vue? Debugger?

Comment (ulfgebhardt) [reply backend/.eslintrc.js]: fixed it tho <3

File: backend/.prettierrc.js (hunk 3 above)
Comment (ulfgebhardt) [reply backend/.prettierrc.js]: Good catch - eslint ignores hidden files by default - do you wish to change this?

File: backend/ormconfig.json (hunk 4 above)
Comment (ulfgebhardt) [reply backend/ormconfig.json:2]: obviously - the package is not even installed

File: backend/package.json (hunk 5 above)
Comment (ulfgebhardt) [reply backend/package.json:5]: `src/index.ts`?
`build/index.js`?

Which one? `build/index.js` is not in the source repo but most likely would be included in a published package

**Comment by Elweyn** [comment]
Hi @ulfgebhardt, @Mogge, @einhornimmond, 

like we discussed in the last meetings I worked on this issue so we can tunnel the Login & Community Server calls over the Apollo Server.
What @ulfgebhardt and @Mogge prepared was clear so I could understand the base structure and continue to implement the different calls.

Now I am a bit concerned about the datas we want to send.
For the moment we use the two Methods apiPost and apiGet that were writen for the frontend, the methods do interpret the response of the servers.
 
The different szenaries for apiPost are:
Error Cases:
- The state is different from 200 the call send an ErrorObject that is given to the response
- The data.state is warning send an ErrorObject that is also returned in the JSON
Response: error is the catched JavaScript object
```
{
success: false,
result: error
}
```

Success:
Response: result is the axio result object
```
{
success: false,
result: result
}
```

So my question do we need the whole axio response object and the whole error object when we go out of the apiPost Method? Does the frontend need this objects to interpret things?
Or can I withdraw the two objects and use a simplified result like:
ErrorJSON
```
{
  success: false,
  result: error.message
}
```
SuccessJSON
```
{
  success: true,
  result: result.data
}
```

**Comment by ulfgebhardt** [comment]
No the whole object is not needed - feel free to structure the graphql response as you wish. Limit the information you send out to what the client needs.
In this case I would always throw an error whenever `success != true` and let graphql handle the error. The thrown error can contain data which could be used by the client if needed.

This way we have an uniform result:
1. Success case - the structured data
2. Failure case - throw an error and use graphql error propagation

Sounds good?

**Comment by Elweyn** [comment]
OK sounds great. So something like the following code would do it:
```
@Query(() => LoginResponse)
async login(@Args() { email, password }: UnsecureLoginArgs): Promise<LoginResponse> {
    email = email.trim().toLowerCase()
    const result = await apiPost(CONFIG.LOGIN_API_URL + 'unsecureLogin', { email, password })

    if (!result.success) {
      throw new Error(result.data)
    }
    
    return new LoginResponse(result.data)
  }
  ```

File: backend/src/graphql/models/Server.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+@ObjectType()
+export class Server extends BaseEntity {
+  constructor(json: any) {
+    super()
+    this.loginServerPath = json.login_server_path
+  }
+
+  @Field(() => String)
+  loginServerPath: string
+}
```
Comment (Mogge) [inline backend/src/graphql/models/Server.ts:16]: What is this for?

File: backend/src/graphql/models/UserInfoData.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+@Entity()
+@ObjectType()
+export class GetUserInfoResponse extends BaseEntity {
+  @Field(() => Boolean)
+  success: boolean
+
+  @Field(() => UserInfoData)
+  data: UserInfoData
+}
```
Comment (Mogge) [inline backend/src/graphql/models/UserInfoData.ts:40]: What is this model for?

File: community_server/src/Controller/AppRequestsController.php [hunk 8]
```diff
@@ -301,12 +303,16 @@ public function getBalance($session_id = 0)
...
+        $this->addAdminError("AppRequests", "getBalance", $body, $user['id']);
         $this->set('body', $body);
     }
     
     public function listTransactions($page = 1, $count = 25, $orderDirection = 'ASC', $session_id = 0)
     {
+        
         $this->viewBuilder()->setLayout('ajax');
+        $this->response = $this->response->withType('application/json');
+
```
Comment (Mogge) [inline community_server/src/Controller/AppRequestsController.php:315]: What have these changes to do with the setup of the apollo server. Do we need them? If so, then why?

File: community_server/src/Template/AppRequests/get_balance.ctp [hunk 9]
```diff
@@ -6,6 +6,10 @@
...
  */
 
-$body['balance'] = $this->element('centToFloat', ['cent' => $body['balance'], 'precision' => 4]);
-$body['decay'] = $this->element('centToFloat', ['cent' => $body['decay'], 'precision' => 4]);
+if(isset($body['balance'])) {
+    $body['balance'] = $this->element('centToFloat', ['cent' => $body['balance'], 'precision' => 4]);
+}
+if(isset($body['decay'])) {
+    $body['decay'] = $this->element('centToFloat', ['cent' => $body['decay'], 'precision' => 4]);
+}
```
Comment (Mogge) [inline community_server/src/Template/AppRequests/get_balance.ctp:14]: Same here? Why do we have changes here?

**Review by Mogge — CHANGES_REQUESTED** [review]
In first place. Why do we have changes on the community server here?

**Review by Mogge — APPROVED** [review]

**Pull request merged**